### PR TITLE
Provide a way to calculate ComposeScene preferredSize that is needed for Swing

### LINKED ISSUE #3834: ComposePanel doesn't calculate its preferredSize correctly when it's called before doLayout
**Describe the bug**
When we have ComposePanel inside e.g. JPanel with BorderLayout it's preferred size will be calculated incorrectly if it is called before doLayout stage.
The reason is that size is equal to 0 at this step and we set such max constraints to the RootOwner.

The same issue can be with iOS interop in future.

**Affected platforms**
Select one of the platforms below:
- Desktop

**Versions**
- Kotlin version*: 1.9.0+
- Compose Multiplatform version*: 1.5.3+
- OS version(s)* (required for Desktop and iOS issues): Windows/Linux/MacOs
- OS architecture (x86 or arm64): All
- JDK (for desktop issues): JBR-17

**To Reproduce**
Try this test

```kotlin
  @Test
  fun `initial panel size with border layout`() {
      assumeFalse(GraphicsEnvironment.getLocalGraphicsEnvironment().isHeadlessInstance)

      runBlocking(MainUIDispatcher) {
          val composePanel = ComposePanel()
          composePanel.setContent {
              Text("Content")
          }

          val frame = JFrame()
          try {
              val content = JPanel(BorderLayout()).apply {
                  add(composePanel, BorderLayout.CENTER)
              }
              frame.contentPane.add(content)
              frame.pack()

              frame.isVisible = true
              delay(1_000)
              assertTrue(content.size.height > 2)
              assertTrue(content.size.width > 2)
          } finally {
              frame.dispose()
          }
      }
  }
```

### PR DESCRIPTION
**Pull request by Walingar**
## Proposed Changes

Previous solution with preferred size didn't work well because constraints with maxSize taken from component.size were used, but component.size is not set (0 by default).

See issue: https://github.com/JetBrains/compose-multiplatform/issues/3834

By Swing design preferred size should be calculated without incoming constraints. It is not possible in Compose, so I artificially set max constraints taking sum of displays dimensions here. Not the best solution, but I couldn't imagine better one..

## Testing

Test: run ComposePanelTest.kt

## Issues Fixed

Fixes: https://github.com/JetBrains/compose-multiplatform/issues/3834

### CODE REVIEW

File: compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/platform/SkiaBasedOwner.skiko.kt
Comment (Walingar) [inline compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/platform/SkiaBasedOwner.skiko.kt:255]: I couldn't find a better way to measure in given constraints...

File: compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/awt/ComposeBridge.desktop.kt
Comment (Walingar) [inline compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/awt/ComposeBridge.desktop.kt]: Comparing to the previous solution, this is much worse from performance side, since before we measured only in constraints no more than given size, and now we always measure in max constraints

File: compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/platform/SkiaBasedOwner.skiko.kt
Comment (Walingar) [reply compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/platform/SkiaBasedOwner.skiko.kt:255]: Also, as I understand it will break layout caches...

File: compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/awt/ComposeBridge.desktop.kt
Comment (m-sasha) [inline compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/awt/ComposeBridge.desktop.kt]: Why not just `Constrains(maxWidth = Constraints.Infinity, maxHeight = Constraints.Infinity)`? These are also the default values, so just `Constraints()` works.

File: compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/awt/ComposeBridge.desktop.kt
Comment (Walingar) [reply compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/awt/ComposeBridge.desktop.kt]: Just thought that we can use max(displays.dimension) here, not sum

File: compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/platform/SkiaBasedOwner.skiko.kt
Comment (Walingar) [reply compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/platform/SkiaBasedOwner.skiko.kt:255]: But it should be called on parent relayout which should happen a lot

File: compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/awt/ComposeBridge.desktop.kt
Comment (m-sasha) [reply compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/awt/ComposeBridge.desktop.kt]: I would do this in a function, since it's not a simple/quick operation.

File: compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/ComposeScene.skiko.kt
Comment (m-sasha) [inline compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/ComposeScene.skiko.kt:485]: We should cache this or turn it into a function. Code that previously assumed this was a quick operation will need to be rewritten. For example in `SemanticsNode.isInScreenBounds()`:

```
    return nodeBoundsInWindow.top >= 0 &&
        nodeBoundsInWindow.left >= 0 &&
        nodeBoundsInWindow.right <= composeView.contentSize.width &&
        nodeBoundsInWindow.bottom <= composeView.contentSize.height
```

Comment (m-sasha) [reply compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/ComposeScene.skiko.kt:485]: Actually, are you sure this isn't breaking anything? Before, contentSize was the actual size of the scene. Now it's the result of some computation - why is it guaranteed to match the real size of the scene?

Comment (Walingar) [reply compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/ComposeScene.skiko.kt:485]: Ah, I missed this particular case, indeed this one can break, thank you

File: compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/awt/ComposeBridge.desktop.kt
Comment (Walingar) [reply compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/awt/ComposeBridge.desktop.kt]: The only reason is the ability to use Lazy* or not. Since when you try to measure e.g. LazyColumn in Infinity height it will throw an exception.
And it won't be so easy to understand what's the reason of the error. 
Also, it can break current applications

Comment (Walingar) [reply compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/awt/ComposeBridge.desktop.kt]: But I am for Inifinity + handle exception, check exception text and provide additional information, that preferred size is calculated -> set preferred size/add specific size for Lazy*

Comment (m-sasha) [reply compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/awt/ComposeBridge.desktop.kt]: Perhaps it's better to say that we don't support putting a layout with undefined intrinsic size as the root element of a `ComposePanel`? Will the result really be better when the `ComposePanel` reports its preferred size to be the entire screen?

Comment (Walingar) [reply compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/awt/ComposeBridge.desktop.kt]: The only thing that I worry about is breaking existing applications. For sure, only not properly written applications will be broken

Comment (m-sasha) [reply compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/awt/ComposeBridge.desktop.kt]: We could measure like before in the case of a fully Compose app (WindowComposeBridge). Or decide it's not a valid case to have a window size of `Dp.Unspecified` and a content with no intrinsic size. @igordmn what do you think?

File: compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/platform/SkiaBasedOwner.skiko.kt
Comment (igordmn) [inline compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/platform/SkiaBasedOwner.skiko.kt:252]: Let's make it public and experimental. The policy for ComposeScene - if something is needed for one platform, it can be needed for other platforms, implemented by users.

Comment (igordmn) [reply compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/platform/SkiaBasedOwner.skiko.kt:252]: Also, let's add TODO for `contentSize`:
```
TODO: investigate all internal usages, and deprecated it, if it isn't actually needed
```

I found one usage in `ui-test`, and I am not sure that is correct usage of `contentSize` there.

File: compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/awt/ComposeBridge.desktop.kt
Comment (igordmn) [reply compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/awt/ComposeBridge.desktop.kt]: I experimented with the alternatives:
1. calling intrinsicSize. Doesn't work, as we have an exception if intrinsic size isn't supported by some widget (LazyList)
2. passing Infinity constraints, as Android does that for WRAP_CONTENT. It works for Android, because Android already knows if the container asks for WRAP_CONTENT or for a fixed size. We can safely throw an exception, if it asks for WRAP_CONTENT, and we have LazyList. It isn't the case with BorderLayout - we don't know if Swing will request WRAP_CONTENT, or will apply fixed constraints afterwards (ADDED: NOT SURE ABOUT THAT). We can't throw an exception.
3. passing zero constraints. Doesn't work, as widgets shrink themselves to zero, without giving any extra information

So, passing some big value to calculate content size is the only option. Another option is to modify the API, but it is harder, so let's keep the current approach, until we encounter issues with it.

File: compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/awt/ComposeBridge.desktop.kt
Comment (igordmn) [inline compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/awt/ComposeBridge.desktop.kt]: Let's use `graphicConfiguration` if it is available as in `Window.setSizeImpl(size: DpSize)`.

Better to use a concrete screen device, if we already know on which screen we show the panel.

File: compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/platform/SkiaBasedOwner.skiko.kt
Comment (igordmn) [inline compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/platform/SkiaBasedOwner.skiko.kt]:
```suggestion
            measureAndLayoutDelegate.measureOnly()
```

We didn't have this function before, but it is more suitable

File: compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/platform/SkiaBasedOwner.skiko.kt
Comment (igordmn) [reply compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/platform/SkiaBasedOwner.skiko.kt:255]: I checked, and Swing doesn't call getPreferredSize on window resize in the `initial panel size of LazyColumn with border layout` test. So, we are safe, and we only have an additional measure at startup?

File: compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/awt/ComposeBridge.desktop.kt
Comment (igordmn) [reply compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/awt/ComposeBridge.desktop.kt]: > We could measure like before in the case of a fully Compose app

We actually measured the same way as in this PR (see `Window.setSizeImpl(size: DpSize)`) for pure Compose applications. Now we use this logic for ComposeWindow/ComposePanel as well.

>  it's not a valid case to have a window size of Dp.Unspecified and a content with no intrinsic size

It is not a valid case, we can't calculate preferred size of LazyList in any meaningful way. 

Actually, I have a question, is there a case in Swing for that? If we don't call `pack()` will `getPreferredSize()` be called? If not, I would just pass Infinity and throw an exception (without handling it).

Comment (Walingar) [reply compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/awt/ComposeBridge.desktop.kt]: `If we don't call pack() will getPreferredSize()` preferredSize can be called anytime see the related issue:  https://github.com/JetBrains/compose-multiplatform/issues/3913

File: compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/platform/SkiaBasedOwner.skiko.kt
Comment (Walingar) [reply compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/platform/SkiaBasedOwner.skiko.kt]: I tried to use this function, but it didn't work for me, since it just reuses previous layout, as I understand. So we anyway need to make a layout

File: compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/platform/SkiaBasedOwner.skiko.kt
Comment (Walingar) [reply compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/platform/SkiaBasedOwner.skiko.kt:255]: It depends on LayoutManager, in case of window resize everything should be fine, since PreferredSize is used only for initial size. But other LayoutManagers may call preferredSize any time :(

File: compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/awt/ComposeBridge.desktop.kt
Comment (igordmn) [reply compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/awt/ComposeBridge.desktop.kt]: This case is valid - we use preferred size to determine the size of popup.

But is there any functionality that calls `getPreferredSize` and ignores its result? If there is no any, we can safely just pass Infinity as constraints, and throw an exception for LazyList

File: compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/platform/SkiaBasedOwner.skiko.kt
Comment (igordmn) [reply compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/platform/SkiaBasedOwner.skiko.kt]: It is not what description says. Layout is a about assignment positions to nodes (after we assigned sizes). And positions don't affect `measuredWidth`. What kind of issues you encountered?

I changed this line, and ComposePanelTest still pass

File: compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/awt/ComposeBridge.desktop.kt
Comment (Walingar) [reply compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/awt/ComposeBridge.desktop.kt]: I'd provide one more usecase then: completion popup with just a few items. It will always be a LazyColumn, but it hcan have thousands or just a few items and when there are only a few of themm then preferredSize should return the exact size, not throwing an exception I guess.
But may be it can be properly handled from the user-side?

Comment (igordmn) [reply compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/awt/ComposeBridge.desktop.kt]: > But may be it can be properly handled from the user-side?

In this case, the user should specify the max size of the lazy list. Ther won't be an exception then.

Comment (Walingar) [reply compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/awt/ComposeBridge.desktop.kt]: Good point, agree with Infinite constraints then

File: compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/ComposeScene.skiko.kt
Comment (Walingar) [reply compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/ComposeScene.skiko.kt:485]: Mentioned code was recently changed by Ivan. Also, this API always was not that cheap, since it did measure and layout.
So, nothing is changed here, code still works the same

File: compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/awt/ComposeBridge.desktop.kt
Comment (Walingar) [reply compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/awt/ComposeBridge.desktop.kt]: Decided to proceed with Infinity constraints

File: compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/platform/SkiaBasedOwner.skiko.kt
Comment (Walingar) [reply compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/platform/SkiaBasedOwner.skiko.kt:252]: It seems that this is not used anymore. Anyway, `contentSize` api is not changed. It still returns the size in mainOwner constraints

File: compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/ComposeScene.skiko.kt
Comment (Walingar) [inline compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/ComposeScene.skiko.kt:485]: Just note, that after my changes this code doesn't perform layout anymore

**Comment by Walingar** [comment]
@igordmn @m-sasha Folks, sorry for the delay. I've answered on comments and pushed recent changes, please take a look.

I've changed max constraints to infinity, so exception will be thrown if user doesn't provide maxSize for lazy layouts.
Also, I provided experimental API in ComposeScene since it is needed by convention of ComposeScene.

It seems that `ComposeScene.contentSize` is not used anymore in our codebase. But since it is public API, I keep it here. The only change is that it doesn't perform layout

**Review by igordmn — APPROVED** [review]

File: compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/ComposeScene.skiko.kt
Comment (igordmn) [inline compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/ComposeScene.skiko.kt]: After we apply infiinity contraints, what do you think, if we:
1. remove `contentSizeInConstraints`
2. place this code into `contentSize`
3. cache `_contentSize` and reset it in `neadLayout`

Comment (Walingar) [reply compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/ComposeScene.skiko.kt]: Sorry, I don't get it. How then can we get contentSize in infinity constraints?

Comment (igordmn) [reply compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/ComposeScene.skiko.kt]: contentSize will set infinity constraints itself. it was added for that in the first place, not to know the content size in the current constraints

Comment (Walingar) [reply compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/ComposeScene.skiko.kt]: Discussed that contentSize should return size in infinity contraints by design. This will be a breaking change, but if someone used this API before, most likely it was used inproperly

Comment (Walingar) [reply compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/ComposeScene.skiko.kt]: Updated PR, please check the new approach

Comment (Walingar) [reply compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/ComposeScene.skiko.kt]: I would like to avoid redundant caching and complicate invalidation, so I just provided calculateContentSize function

File: compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/ComposeScene.skiko.kt
Comment (igordmn) [inline compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/ComposeScene.skiko.kt:502]: Let's inline the content into `calculateContentSize`. A separate function doesn't add any value now

File: compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/ComposeScene.skiko.kt
Comment (igordmn) [inline compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/ComposeScene.skiko.kt]:
```suggestion
    @Deprecated("Use calculateContentSize() instead", replaceWith = ReplaceWith("calculateContentSize()"))
```
We lean towards to not removing deprecated API until 2.0

Comment (igordmn) [reply compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/ComposeScene.skiko.kt]: And better to have `calculateContentSize` experimental for now, with `TODO: makes stable and deprecate contentSize after 1.6`

**Review by igordmn — APPROVED** [review]
LGTM. Let's wait for the second approval from @m-sasha

**Comment by m-sasha** [comment]
Don't know how important this is, but some of the javadocs exceed 100 columns.

File: compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/ComposeScene.skiko.kt
Comment (m-sasha) [inline compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/ComposeScene.skiko.kt:496]: Maybe call it `preferredSize()`?

File: compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/ComposeScene.skiko.kt
Comment (m-sasha) [inline compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/ComposeScene.skiko.kt:481]: It seems to me that "current content size" and "preferred size" (in infinite constraints) are not at all the same thing. If so, why do we want to deprecate getting the current size?

File: compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/platform/SkiaBasedOwner.skiko.kt
Comment (m-sasha) [reply compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/platform/SkiaBasedOwner.skiko.kt:252]: @igordmn `contentSize` is public API, and `calculateContentSize` is not a replacement, as it does something else entirely.

File: compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/ComposeScene.skiko.kt
Comment (Walingar) [reply compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/ComposeScene.skiko.kt:481]: @igordmn told that by design contentSize was expected to provide size in infinity constraints, so there is no usecase to get current size and should be received in another way

**Review by m-sasha — APPROVED** [review]

File: compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/ComposeScene.skiko.kt
Comment (igordmn) [reply compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/ComposeScene.skiko.kt:481]: Yes, there wasn't an intention to provide current content size, and I am not aware of use cases where we need it. It was added to determine the size of the window/panel before its construction.

File: compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/ImageComposeScene.skikoMain.kt
Comment (MatkovIvan) [inline compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/ImageComposeScene.skikoMain.kt:205]: JFYI I'll mark this as `ExperimentalComposeUiApi` too in my branch

Comment (igordmn) [reply compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/ImageComposeScene.skikoMain.kt:205]: If we mark the whole ComposeScene as `@Internal`, then there is no need for that (mark calculateContentSize)

Comment (MatkovIvan) [reply compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/ImageComposeScene.skikoMain.kt:205]: @igordmn It's `ImageComposeScene`

Comment (igordmn) [reply compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/ImageComposeScene.skikoMain.kt:205]: Ok, make sense there

**Pull request merged**
